### ISSUE TITLE
Reordering blanks: problems with negative numbers

### ISSUE THREAD

**Issue by hectoralos** [apertium/apertium#107]
We currently get this:

```
$ echo "En -407" | apertium -d . fra-cat
El -any 407
```

It should be `L'any -407`, but the hyphen has been reordered.

**Comment by khannatanmai** [comment]
@hectoralos I would say that the - sign should be analysed with the number.

**Comment by khannatanmai** [comment]
Or analyse - as a token and then combine it with the number in separable.

**Comment by khannatanmai** [comment]
In the french dix:
```
<pardef n="numeros">
  <e>       <re>[0-9]+([.,][0-9]+)?</re><p><l></l><r><s n="num"/></r></p></e>
  <e>       <re>[1-9][0-9]+</re><p><l>e</l><r>ème<s n="adj"/><s n="mf"/><s n="sg"/></r></p></e>
  <e>       <re>[1-9][0-9]+</re><p><l>es</l><r>ème<s n="adj"/><s n="mf"/><s n="pl"/></r></p></e>
  <e r="LR"><re>[1-9][0-9]+</re><p><l>ème</l><r>ème<s n="adj"/><s n="mf"/><s n="sg"/></r></p></e>
  <e r="LR"><re>[1-9][0-9]+</re><p><l>èmes</l><r>ème<s n="adj"/><s n="mf"/><s n="pl"/></r></p></e>
<!--  <e>       <re>[0-9]([0-9]([0-9])?)?( [0-9][0-9][0-9])*</re><p><l></l><r><s n="num"/></r></p></e> -->
<!--  <e>       <re>[0-9]([0-9]([0-9])?)?( [0-9][0-9][0-9])*([.,][0-9]+)?</re><p><l></l><r><s n="num"/></r></p></e> -->
</pardef>
```

Can change the regex `` [0-9]+([.,][0-9]+)?`` to ``-?[0-9]+([.,][0-9]+)?``.

This would mean that if the input has ``10-20`` then it gets analysed as`` ^10<num>$^-20<num>$``, but it shouldn't ever be reordered so I would say it should be fine, but can discuss it. Can also change the regex such that the hyphen gets analysed as part of the number if the hyphen doesn't immediately follow a number.

**Comment by hectoralos** [comment]
Yes, I imagine how to do it, but this will have to be done in all the ~50 language pairs. I suspect that more than a half of them don't have anyone who maintains them. Probably, guessing that there is a negative number in the text will be useful for most of the language pairs.

**Comment by khannatanmai** [comment]
@hectoralos yeah those can be changed when the maintainers need to change it, and if we agree on this strategy about negative numbers then I'm sure we can just change it in all the pairs in one go.

**Comment by hectoralos** [comment]
The inhability of having backward compatibility with the new blank-reordering stuff is a huge problem. Yes, probably something automatic can be done for supporting negative numbers in the dictionaries. Another very different thing is the possibility of automatically creating new versions of language pairs and release them. Maybe this can work for two-letter pairs, but it will not for language pairs that share language packages. For instance, "my" Catalan-Sardinian pair will need work to pass testvoc because the Sardinian package has undergone changes due to a new version of the Italian-Sardinian package, on which we are currently working. So, I really prefer a clarification from the PMC about the strategy of deploying the new Apertium version. Seemingly, it will solve a core problem in Apertium, but it has its costs.

**Comment by khannatanmai** [comment]
So, the blank reordering stuff that was done with wordbound blanks is fully backwards compatible. What wasn't backwards compatible was the changes made to clipping blanks in transfer. Because blanks were clipped using position, sometimes blanks that existed in the source would not appear in the target, which was a massive problem, as it often gave different number of lines in the output than the input. Secondly, the position capability was given to reorder blanks but now with wordbound blanks, reordering blanks was no longer needed. That's why after consulting with several people I decided to make it so blanks don't reorder anymore and if there's a blank in the input it is guaranteed to appear in the output.

The issue started when I found out people are using blank reordering to deal with punctuations, which is something that was going to have to be fixed anyway, so that's the part that needs to be changed now. It's a small price to pay for the formatting issues people were facing for years, and I'm not even talking about html files, but just that a translator should give the same lines in the output as input, otherwise post-editing becomes a pain.

**Comment by unhammer** [comment]
> For instance, "my" Catalan-Sardinian pair will need work to pass testvoc because the Sardinian package has undergone changes due to a new version of the Italian-Sardinian package, on which we are currently working.

@TinoDidriksen  can clarify, but I don't think you need to do a new cat-srd release just because srd got a new release; the old cat-srd package will use the old srd package and the new ita-srd will use the new srd. Of course, when you're ready to do a new cat-srd release, you'll have to do testvoc, but that's as it has always been.

**Comment by unhammer** [comment]
> Or analyse - as a token and then combine it with the number in separable.

That's how nno-nob has always done it (`^-/-<guio>$`), though `^-407/…$` would probably be better. Both solutions work and are not much trouble to implement. The minus/dash is such a frequent and important character that I don't see why it should not have an analysis.

**Comment by hectoralos** [comment]
> people are using blank reordering to deal with punctuations

Is really someone doing this? I wasn't aware. It seems very strange to me. What happens is that some signs are considered as part of the blanks in many languages. As we discussed, in some cases this is the easier solution, but in others it does not work. Probably quotation marks would have to be treated as punctuation, but this implies a lot of work for disambiguation and transfer. Currently nor one not the other support this for at least several core Apertium languages, like Spanish and Catalan. So, this again comes to the problem of the deployment of the new version of Apertium. I don't dispute that the new version is a great work and, as I said, it solves a core problem in Apertium, but its releasing without changing many language pairs has serious implications, as this issue shows. That's why I think the PMC should give clear criteria and maybe a calendar about what to do.

**Comment by hectoralos** [comment]
> The minus/dash is such a frequent and important character that I don't see why it should not have an analysis.

@unhammer , Well, I found this when I begin to work with Apertium, but I can very well understand why this have been done this way. If you have *this "strange" noun clause* in English and you have to translate it into Spanish, you will have to add quite a lot of possible NC dealing with whether to determinant, the adjective, the noun or a combination of them have quotation marks. Otherwise, your translation will not produce a correct translation that changes the order of the adjective and the noun and generates the apropriate gender coordination. At this is just the case for one possible NC. There are quite a lot of them. A similar problem happens in disambiguation.

**Comment by hectoralos** [comment]
> > For instance, "my" Catalan-Sardinian pair will need work to pass testvoc because the Sardinian package has undergone changes due to a new version of the Italian-Sardinian package, on which we are currently working.
> 
> @TinoDidriksen can clarify, but I don't think you need to do a new cat-srd release just because srd got a new release; the old cat-srd package will use the old srd package and the new ita-srd will use the new srd. Of course, when you're ready to do a new cat-srd release, you'll have to do testvoc, but that's as it has always been.

Yes, I know. What I was trying to said is that I am afraid we'll have to make new express versions of the pairs. In this case, there will be pairs that will need more work because they are not as up-to-date as others.

**Comment by ftyers** [comment]
> PMC should give clear criteria and maybe a calendar

I think having a calendar for changeover is an excellent idea.

**Comment by TinoDidriksen** [comment]
> That's why I think the PMC should give clear criteria and maybe a calendar about what to do.

A calendar is not really a thing we can do. It's done. It's already live for all development work. It's not hidden away in a branch, nor can or should it be. Development should be done with latest tools, precisely so that these issues get noticed, and, where possible, globally fixed.

As for criteria, I would say: Analyze everything. This is what we eventually had to do for GramTrans / VISL as well. We used to do as Apertium does and only analyze words and some punctuation, but the same issues cropped up when we started handling more complex formats. Analyzing everything is absolutely a big change and will require a lot of work, but it brings with it more correct analyses and better translations. Symbols have meaning, especially things like quotes. [CG has a cool feature if you analyze quotes](https://visl.sdu.dk/cg3/chunked/parentheses.html).

As Giellatekno does, we should have a shared source of analyses for symbols and other cross-language patterns (URLs, emails, etc).

**Comment by mr-martian** [comment]
The point has been brought up several times that analyzing punctuation significantly increases the number of patterns. It seems to me that adding shortcuts for common patterns would mitigate this problem somewhat and is something that I would be happy to do.

**Comment by hectoralos** [comment]
@mr-martian, sounds useful. What would you propose more specifically?

**Comment by mr-martian** [comment]
One thing in particular that comes to mind on the subject of punctuation is in -recursive to have something like
```
QUOTES: (@lpar )@rpar, '@apos '@apos, "@quot "@quot;
QUOTED: n, NP;
```
which would then expand into something along the lines of
```
n -> (@lpar n )@rpar { 1 2 3 };
n -> '@apos n 'apos { 1 2 3 };
n -> "@quot n "@quot { 1 2 3 };
NP -> (@lpar NP )@rpar { 1 2 3 };
NP -> '@apos NP 'apos { 1 2 3 };
NP -> "@quot NP "@quot { 1 2 3 };
```
that is, take a list of pair punctuation and a list of constituents and generate rules which match that constituent between those punctuation marks and generates a chunk with the same tags (which effectively just ignores the punctuation).

In modules that are not transfer (lexical selection, maybe -separable), I think it would be possible to have a way of marking rules as `skippunct=true` or something like that.

More generally though, I consider "I wish there was an easier way to do X" to be a perfectly reasonable feature request. It also seems a bit more likely to produce useful abbreviations than me trying to predict in advance what will be useful.

**Comment by khannatanmai** [comment]
If we really need the ability to reorder blanks, we could take a step back and add the <b pos="X"/> feature again, but keep the automatic flushing of blanks.

**Comment by TinoDidriksen** [comment]
> If we really need the ability to reorder blanks...

Erk, no. Re-ordering normal blanks is the wrong thing to do. The very fact that one would want them re-ordered means there is semantic relevance, which means it should've been analyzed.

**Comment by hectoralos** [comment]
@mr-martian , I don't understand exactly your pseudo-code, but I can broadly imagine the meaning. The idea seems that the current rules will work, but e.g. "n" would not be only "n" but "quot + n", "n + quot", "quot + n + quot", etc. if a flag is raised. If feasible, it looks promising. This eventually could be generalized, for instance, for adv + adj, preadv + adj as synonyms of adj (but, depending on the language, the order should be the opposite). What do you think, @jaumeortola , @xavivars ?

**Comment by jonorthwash** [comment]
@hectoralos, could you clarify whether you're using shallow or recursive transfer?

**Comment by hectoralos** [comment]
I've not yet used recursive transfer.

**Comment by jonorthwash** [comment]
> I've not yet used recursive transfer.

Ah, that would explain the confusion.  The code that @mr-martian posted wasn't pseudo-code, but the actual `apertium-recursive` formalism.
